sdk/core/core-lro/test/testClient.test.ts

```diff
@@ -0,0 +1,263 @@
+    client.setResponses([initialResponse, ...Array(20).fill(basicResponseStructure)]);
+
+    const poller = await client.startNonCancellableLRO();
+    poller.done().catch((e) => {
+      assert.equal(e.message, "Poller stopped");
+    });
+
+    let operation = poller.operation;
+    assert.ok(operation.state.started);
+    assert.equal(client.totalSentRequests, 1);
+
+    await delay(100); // The poller loops every 10 milliseconds
+    assert.equal(client.totalSentRequests, 10);
+
+    let error: any;
+    try {
+      await poller.cancel();
+    } catch (e) {
+      error = e;
+    }
+    assert.equal(error.message, "Cancellation not supported");
+    poller.stop();
+  });
+
+  it("can stop polling the operation", async function() {
+    const client = new TestClient(new SimpleTokenCredential("my-test-token"));
+    client.setResponses([initialResponse, ...Array(20).fill(basicResponseStructure)]);
+
+    const poller = await client.startLRO();
+    poller.done().catch((e) => {
+      assert.equal(e.message, "Poller stopped");
+    });
+
+    let operation = poller.operation;
+    assert.ok(operation.state.started);
+    assert.equal(client.totalSentRequests, 1);
+
+    await delay(100); // The poller loops every 10 milliseconds
+    assert.equal(client.totalSentRequests, 10);
+
+    poller.stop();
+
+    await delay(100);
+    assert.equal(client.totalSentRequests, 10);
+  });
+
+  it("can document progress", async function() {
+    const client = new TestClient(new SimpleTokenCredential("my-test-token"));
+    client.setResponses([
+      initialResponse,
+      ...Array(10).fill(basicResponseStructure),
+      doFinalResponse,
+      finalResponse
+    ]);
+
+    const poller = await client.startLRO();
+
+    let totalOperationUpdates = 0;
+    let lastOperationUpdate: TestOperation = poller.operation;
+    poller.onProgress((operation) => {
+      totalOperationUpdates++;
+      lastOperationUpdate = operation;
+    });
+
+    const result = await poller.done();
+    assert.equal(result, "Done");
+
+    assert.equal(totalOperationUpdates, 13);
+
+    const { properties, state } = lastOperationUpdate;
+    assert.ok(properties.initialResponse!.parsedBody.started);
+    assert.ok(properties.previousResponse!.parsedBody.finished);
+    assert.ok(state.completed);
+  });
+
+  it("can reuse one poller state to instantiate another poller", async function() {
+    const client = new TestClient(new SimpleTokenCredential("my-test-token"));
+
+    // A total of 13 expected responses.
+    const responses = [
+      initialResponse,
+      ...Array(10).fill(basicResponseStructure),
+      doFinalResponse,
+      finalResponse
+    ];
+    client.setResponses(responses);
+
+    const poller = await client.startLRO();
+    poller.done().catch((e) => {
+      assert.equal(e.message, "Poller stopped");
+    });
+
+    await delay(100); // The poller loops every 10 milliseconds
```

Comment:
Since you did the work to make delay abstract, why not make the test poller have a public method to trigger the completion of `delay` so you don't have to coordinate timeouts and take real world clock time?

Reply:
I would like to not do this for the tests, since my internal delays are representative of 1/10 of what I'm trying to measure here. Is it a problem if we keep using a greater delay during the tests? to compare against several iterations of the internal delay?

Reply:
what if your test is running on a cheap CI VM in the cloud and it takes longer than 10ms for your callback to be called? 

Relying on things to happen in real-world time means sometimes they don't happen that way because of scheduling/resource problems. Maybe I'm overindexing here, but coming from a world with a ton of flaky tests due to CI timing issues, it seems good to avoid needing to count on your callbacks being fired correctly if we don't have to.